retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/RecordingSubscriber.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package retrofit2.adapter.rxjava;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import rx.Notification;
+import rx.Subscriber;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** A test {@link Subscriber} and JUnit which guarantees all events are asserted. */
+final class RecordingSubscriber<T> extends Subscriber<T> {
+  private final long initialRequest;
+  private final Deque<Notification<T>> events = new ArrayDeque<>();
+
+  private RecordingSubscriber(long initialRequest) {
+    this.initialRequest = initialRequest;
+  }
+
+  @Override public void onStart() {
+    request(initialRequest);
+  }
+
+  @Override public void onNext(T value) {
+    events.add(Notification.createOnNext(value));
+  }
+
+  @Override public void onCompleted() {
+    events.add(Notification.<T>createOnCompleted());
+  }
+
+  @Override public void onError(Throwable e) {
+    events.add(Notification.<T>createOnError(e));
+  }
+
+  private Notification<T> takeNotification() {
+    Notification<T> notification = events.pollFirst();
+    if (notification == null) {
+      throw new AssertionError("No event found!");
+    }
+    return notification;
+  }
+
+  public T takeValue() {
+    Notification<T> notification = takeNotification();
+    assertThat(notification.isOnNext())
+        .overridingErrorMessage("Expected onNext event but was %s", notification)
```

Comment:
This isn't a reusable utility so I don't feel compelled to bother. The
error message already makes it really easy to understand what went wrong
since you can only assert one thing at a time and the stacktrace indicates
which assertion failed.

On Thu, Aug 18, 2016 at 11:38 AM Artem Zinnatullin <notifications@github.com>
wrote:

> In
> retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/RecordingSubscriber.java
> <https://github.com/square/retrofit/pull/1980#discussion_r75332492>:
>
> > +  @Override public void onError(Throwable e) {
> > +    events.add(Notification.<T>createOnError(e));
> > +  }
> > +
> > +  private Notification<T> takeNotification() {
> > +    Notification<T> notification = events.pollFirst();
> > +    if (notification == null) {
> > +      throw new AssertionError("No event found!");
> > +    }
> > +    return notification;
> > +  }
> > +
> > +  public T takeValue() {
> > +    Notification<T> notification = takeNotification();
> > +    assertThat(notification.isOnNext())
> > +        .overridingErrorMessage("Expected onNext event but was %s", notification)
>
> To easily see actual state of subscriber without diving into details of
> test case code, like:
>
> Expected onNext event but was onError.
>
> State: onNext("a"), onNext("b"), onError("IllegalStateException(whoa)")
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/square/retrofit/pull/1980/files/efc0dd57c3fd6abefe1f4717840d7ea8f886b929#r75332492>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEEEXOGGNGuw_YefDo7EDinFL4evOWEks5qhHyAgaJpZM4Jnhl6>
> .
>